### DISCUSSION TITLE
Injecting request/response into services, repositories or other container managed entities.

### DISCUSSION THREAD

**Discussion by roberttaylortech** [loopbackio/loopback-next#8107] [Q&A] [answered]
Super busy trying to ship something using Loopback 4... and haven't really been able to grok the code or step through to understand everything about the DI system. However.. I do need quick answer and help *please*! :)

I've run into the a few times in a few spots... how can we inject `@inject(RestBindings.Http.REQUEST) protected req: Request;` the request and response more flexibly in more in more places. We tried a couple of spots - first a Service and now a Repository (extends DefaultCrudRepository) and gotten the error below. Injecting req/res into controllers works great.... how can we achieve 
more flexible req/res injection in other component types?

`Cannot start the application. ResolutionError: The key 'rest.http.request' is not bound to any value in context RestApiApplication-jqw2cI9OSNK78Mr8SBCTDQ-0 (context: SRestApiApplication-jqw2cI9OSNK78Mr8SBCTDQ-0, binding: rest.http.request, resolutionPath: repositories.EntitlementRepository -`

Help is much appreciated! A bit of an urgent need on this!

**Comment by raymondfeng** [comment]
The request/response object can only be injected to a controller/service/... with `REQUEST` binding scope.

**Comment by achrinza** [comment, marked as answer]
It seems that this is happening at startup. It works on Controllers as they are instantiated on a per-request basis, meaning that the binding is already populated by the time the Controller instance is created. In contrast, Services, Components, etc. are instantiated on startup by default. This means that no HTTP request is active at the time the binding is being resolved, hence why it's throwing that error.

A quick fix would be to use [Getters](https://loopback.io/doc/en/lb4/Decorators_inject.html#injectgetter):

```typescript
// Imports
import {Getter, inject} from '@loopback/rest';
import {Request, RestBindings} from '@loopback/rest';
...
// Constructor / Property / Method dependency injection
@inject.getter(RestBindings.Http.REQUEST) protected reqGetter: Getter<Request>;
...
// Inside an async function that's called only during a HTTP Request-Response cycle
const req = await reqGetter()
```

A more long-term fix would be to reconsider not passing the raw request object to the server/component/etc., and instead have the Controller filter and pass only what's necessary.
